## Sample Project: Analyzing Restaurant Tips 

### **Introduction**  
The restaurant industry often analyzes tipping behavior to optimize customer service and revenue. In this project, we will use the [**tips dataset**](https://github.com/mwaskom/seaborn-data/blob/master/tips.csv) to explore the relationship between various factors, such as meal time, day of the week, and customer demographics, and their impact on tipping behavior.  

### **Objectives**  
1. Understand the distribution of tips and total bills.  
2. Analyze tipping behavior by demographic factors like gender and group size.  
3. Examine how tipping varies across different days and meal times.  
4. Identify patterns or insights that can inform restaurant management strategies.  

---

### 1. Loading and Understanding the Dataset

#### 1.1 Load the Dataset
First, we load the `tips` dataset using seaborn and perform an initial exploration to understand its structure and features.  

The dataset contains the following columns:  
- **total_bill**: Total bill in USD.  
- **tip**: Tip amount in USD.  
- **sex**: Gender of the bill payer.  
- **smoker**: Indicates if the group included smokers.  
- **day**: Day of the week.  
- **time**: Meal type (Lunch/Dinner).  
- **size**: Number of people in the dining group.  


In [14]:
import pandas as pd
import seaborn as sns

tips = sns.load_dataset('tips')       # Load the dataset
# print(tips.head(10))                    # Display the first five rows
tips.describe()

total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000

#### 1.2 Understanding the Dataset Structure

Using the `info()` method, we can quickly examine the structure of the tips dataset. It reveals that the dataset contains 244 entries and 7 columns, all of which have no missing values. The columns include numerical data (`total_bill`, `tip`, `size`) and categorical data (`sex`, `smoker`, `day`, `time`)

. This summary helps us confirm that the data is complete and appropriately typed for analysis. For instance, categorical variables like `sex` and `day` are stored as `category`, which is memory efficient and facilitates operations like grouping and aggregation.


In [5]:
print(tips.info())                   # Summary of the dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   total_bill  244 non-null    float64 
 1   tip         244 non-null    float64 
 2   sex         244 non-null    category
 3   smoker      244 non-null    category
 4   day         244 non-null    category
 5   time        244 non-null    category
 6   size        244 non-null    int64   
dtypes: category(4), float64(2), int64(1)
memory usage: 7.4 KB
None


#### 1.3 Statistical Overview of the Dataset

The `describe()` method provides a statistical summary of the numerical columns in the dataset, such as `total_bill`, `tip`, and `size`. It includes metrics like the mean, standard deviation, minimum, and maximum values, along with the 25th, 50th (median), and 75th percentiles. 

This overview is valuable for understanding the distribution and variability of the data. For example, we can see that the average total bill is around $\$19.79$, with tips averaging about $\$2.99$. This summary helps identify potential outliers, skewness, or unusual patterns, guiding our next steps in data exploration or cleaning.


In [7]:
print(tips.describe())               # Statistical overview

       total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000




---


### 2. Understanding Tip Distributions

To establish a baseline, let’s explore the distribution of tips and the relationship between tips and total bills.

#### 2.1 Summary Statistics for Tips

By using the `describe()` method on the `tip` column, we obtain a detailed summary of the tip amounts in the dataset. The mean tip is 2.99, with a standard deviation of 1.38. The minimum tip is 1.00, and the maximum tip is 10.00. The quartiles show that 50% of tips fall between 2.00 and 3.00, with a median tip of 2.89. This summary helps us understand the typical tipping behavior and identify potential outliers, providing valuable insights for further analysis.
.
.



In [10]:
print(tips['tip'].describe())         # Summary statistics for tips

count    244.000000
mean       2.998279
std        1.383638
min        1.000000
25%        2.000000
50%        2.900000
75%        3.562500
max       10.000000
Name: tip, dtype: float64


#### 2.2 Identifying Unusually High Tips

By filtering the `tip` column using the 95th percentile (`quantile(0.95)`), we can identify unusually high tips. This operation returns the rows where the tip exceeds the value at the 95th percentile, which helps us spot outliers or exceptional tipping behavior. In this case, we observe tips above 6.00, which are higher than the majority of tips in the dataset. Analyzing these extreme values can provide insights into specific customer behaviors or special circumstances, such as large groups or extraordinary service, that may warrant further exploration.
.


In [11]:
print(tips[tips['tip'] > tips['tip'].quantile(0.95)])             # Identify unusually high or low tips

     total_bill    tip     sex smoker   day    time  size
23        39.42   7.58    Male     No   Sat  Dinner     4
44        30.40   5.60    Male     No   Sun  Dinner     4
47        32.40   6.00    Male     No   Sun  Dinner     4
52        34.81   5.20  Female     No   Sun  Dinner     4
59        48.27   6.73    Male     No   Sat  Dinner     4
88        24.71   5.85    Male     No  Thur   Lunch     2
141       34.30   6.70    Male     No  Thur   Lunch     6
170       50.81  10.00    Male    Yes   Sat  Dinner     3
181       23.33   5.65    Male    Yes   Sun  Dinner     2
183       23.17   6.50    Male    Yes   Sun  Dinner     4
212       48.33   9.00    Male     No   Sat  Dinner     4
214       28.17   6.50  Female    Yes   Sat  Dinner     3
239       29.03   5.92    Male     No   Sat  Dinner     3


**Key Insights:**  
- The majority of tips fall within a specific range (e.g., $2–$5).  
- Outliers or unusually high tips may represent exceptional service or large groups.  


#### 2.3 Correlation Between Total Bill and Tip

To explore the relationship between the total bill and the tip, we calculate the correlation coefficient using the `corr()` method. The resulting correlation value of 0.68 indicates a moderate positive correlation, meaning that as the total bill increases, the tip also tends to increase. 

This is expected, as larger bills generally result in higher tip amounts. Understanding this correlation helps us quantify the relationship and can guide business decisions, such as adjusting tipping policies or analyzing how total bill amounts influence overall revenue.


In [22]:
# Correlation between total bill and tip
correlation = tips['total_bill'].corr(tips['tip'])
print(f"Correlation between total bill and tip: {correlation:.2f}")

Correlation between total bill and tip: 0.68


A strong positive correlation indicates that higher bills generally lead to higher tips.

---

### **3: Analyzing Demographic Factors**

#### 3.1 Tipping Behavior by Gender


**Goal:**  
- Compare tipping behavior between male and female customers.  
- Assess if there are significant differences in tipping practices.


Using the `groupby()` method, we calculated the average tip for each gender by grouping the data based on the `sex` column. The results show that male customers, on average, tip 3.09, while female customers tip 2.83

. This indicates a slight difference in tipping behavior between genders, with males tipping slightly more on average. Such insights could be useful for understanding customer behavior and might inform service strategies or further demographic analyses related to tipping patterns.
.


In [26]:
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [38]:
# Average tip by gender
gender_tips = tips.groupby('sex', observed=True)['tip'].mean()
print(gender_tips)

sex
Male      3.089618
Female    2.833448
Name: tip, dtype: float64


#### 3.2 Influence of Group Size on Tips  


By calculating the tip per person (i.e., dividing the total tip by the group size), we can analyze how tipping behavior changes as the group size increases


By calculating the tip per person (i.e., dividing the total tip by the group size), we can analyze how tipping behavior changes as the group size increases. The results show that smaller groups tend to tip more per person, with a tip of 1.44 for a group of 1 person, which gradually decreases as the group size grows. For example, a group of 5 people tips an average of 0.81 per person. This trend suggests that larger groups may contribute a smaller tip per individual, which could be useful when considering automatic gratuities or adjusting service strategies for different group sizes.
.



In [40]:
# Average tip per person by group size
tips['tip_per_person'] = tips['tip'] / tips['size']
group_size_analysis = tips.groupby('size')['tip_per_person'].mean()
print(group_size_analysis)

size
1    1.437500
2    1.291154
3    1.131053
4    1.033851
5    0.805600
6    0.870833
Name: tip_per_person, dtype: float64


In [42]:
tips.head()

total_bill   tip     sex smoker  day    time  size  tip_per_person
0       16.99  1.01  Female     No  Sun  Dinner     2        0.505000
1       10.34  1.66    Male     No  Sun  Dinner     3        0.553333
2       21.01  3.50    Male     No  Sun  Dinner     3        1.166667
3       23.68  3.31    Male     No  Sun  Dinner     2        1.655000
4       24.59  3.61  Female     No  Sun  Dinner     4        0.902500

---

### **4: Temporal Analysis**


#### 4.1 Tipping Patterns by Day of the Week  

**Goal:**  
- Identify which days generate the highest average tips.  
- This could inform staffing or promotional decisions.


Using the `groupby()` method, we calculated the average tip for each day of the week. The results show that tips are highest on Sunday (3.26), followed by Saturday (2.99), with Thursday and Friday being slightly lower (2.77 and 2.73, respectively)ue())


In [21]:
# Average tip by day
tips_by_day = tips.groupby('day', observed=True)['tip'].mean()
print(tips_by_day)

day
Thur    2.771452
Fri     2.734737
Sat     2.993103
Sun     3.255132
Name: tip, dtype: float64


This indicates that customers tend to tip more generously on the weekend. To ensure the dataset only contains data for these four days, we can check for any other days that might be present:


In [64]:
# Check for any other days in the dataset
print(tips['day'].unique())


['Sun', 'Sat', 'Thur', 'Fri']
Categories (4, object): ['Thur', 'Fri', 'Sat', 'Sun']


#### 4.2 Tipping Behavior During Lunch vs. Dinner  

### Average Tip by Time of Day

By grouping the dataset by the `time` column, we calculated the average tip for lunch and dinner. The results show that tips are higher during dinner (average tip of 106) compared to lunch (average tip of 2.73). This suggests that dinner meals, which typically involve larger bills, tend to generate higher tips. Understanding these trends can help restaurants optimize staffing or adjust pricing strategies based on meal times to maximize revenue and improve customer service.


In [23]:
# Average tip by time of day
tips_by_time = tips.groupby('time', observed=True)['tip'].mean()
print(tips_by_time)

time
Lunch     2.728088
Dinner    3.102670
Name: tip, dtype: float64


---

### **5: Advanced Explorations**

#### 5.1 Impact of Smoking on Tips  


By grouping the dataset by the `smoker` column, we calculated the average tip for smoker and non-smoker tables. The results show that the average tip for smoker tables is 3.01, while non-smoker tables average 2.99. Although the difference is very small, smoker tables tend to tip slightly more on average. This subtle trend could provide useful insights for restaurant seating strategies or customer service adjustments, but further analysis would be needed to confirm any significant patterns or underlying factors influencing tipping behavior.
s.


In [26]:
# Compare tips for smoker vs. non-smoker tables
smoker_tips = tips.groupby('smoker', observed=True)['tip'].mean()
print(smoker_tips)

smoker
Yes    3.008710
No     2.991854
Name: tip, dtype: float64


#### 5.2 Percentage Tips  


To better understand tipping behavior, we calculated the tip as a percentage of the total bill using the formula `(tip / total_bill) * 100`. The results show that, on average, males tip 15.77% of their total bill, while females tip 16.65%. This indicates that, on average, female customers leave a slightly higher percentage tip than male customers. These insights could be useful for further analyzing customer behavior and refining restaurant policies or promotional efforts to encourage higher tip percentages across different demographics.


In [28]:
# Calculate tip percentage
tips['tip_percentage'] = (tips['tip'] / tips['total_bill']) * 100

# Average tip percentage by demographic
tip_pct_by_gender = tips.groupby('sex', observed=True)['tip_percentage'].mean()
print(tip_pct_by_gender)

sex
Male      15.765055
Female    16.649074
Name: tip_percentage, dtype: float64


---


### 6. Key Insights from Tipping Analysis


1. **Peak Tipping Times**: Dinner and weekends generate higher tips on average. Dinner tips are generally higher (average tip of 3.06) compared to lunch (average tip of 2.73), and Sundays see the highest tips (average tip of 3.26). Restaurants should ensure experienced staff are available during these peak periods to optimize service and enhance customer satisfaction.

2. **Demographics**: There is a small difference in tipping behavior based on gender, with females tipping slightly more (16.65%) than males (15.77%). Additionally, smoker and non-smoker tables show minimal differences in tipping, with smokers tipping slightly more (3.01) than non-smokers (2.99). These trends can guide marketing strategies or customer service customization, for example, by offering personalized promotions to specific groups.

3. **Group Size Considerations**: Larger groups tend to tip less per person. For instance, group sizes of 5 or more show a decrease in tip per person. Introducing automatic gratuities for larger groups could help ensure that servers receive fair compensation, especially during busy times.

4. **Education on Tipping Norms**: The calculation of tip percentages (average of 15.77% for males and 16.65% for females) suggests that standard tipping practices could be encouraged. Informing customers about typical tipping percentages, such as through menu or receipt notifications, may help ensure more consistent and appropriate tipping behavior.

Through this analysis, we uncovered actionable insights into tipping behaviors. By leveraging Pandas, we efficiently explored and analyzed the dataset, showcasing its powerful capabilities for data manipulation and analysis. Further analyses with visualization libraries like matplotlib or seaborn could strengthen these findings and offer deeper insights.


